iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp

```diff
@@ -0,0 +1,343 @@
+  mlir::iree_compiler::IREE::Codegen::LoweringConfigAttr loweringConfig;
+  if (tilingLevel != -1 &&
+      succeeded(getComputeOps(funcOp, computeOps, tiledLoops))) {
+    for (auto op : computeOps) {
+      if (auto currLoweringConfig = iree_compiler::getLoweringConfig(op)) {
+        if (loweringConfig) {
+          return {funcOp.emitOpError(
+              "unhandled multiple lowering configurations in compute ops")};
+        }
+        loweringConfig = currLoweringConfig;
+      }
+    }
+  }
+  if (!loweringConfig) {
+    return LinalgTilingAndFusionOptions();
+  }
+  LinalgTilingAndFusionOptions options;
+  options.tileSizes.assign(loweringConfig.getTileSizeVals(tilingLevel));
+  return options;
+}
+
+//===----------------------------------------------------------------------===//
+// From Sandbox
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LinalgFusePass : public LinalgFuseBase<LinalgFusePass> {
+  LinalgFusePass(int64_t tilingLevel = -1) {
+    this->tilingLevel.setValue(tilingLevel);
+  }
+  LinalgFusePass(const LinalgFusePass &pass) {}
+  void runOnOperation() override;
+};
+
+struct LinalgSingleTilingExpertPass
+    : public LinalgSingleTilingExpertBase<LinalgSingleTilingExpertPass> {
+  LinalgSingleTilingExpertPass(int64_t tilingLevel = -1) {
+    this->tilingLevel.setValue(tilingLevel);
+  }
+  LinalgSingleTilingExpertPass(const LinalgSingleTilingExpertPass &pass) {}
+
+  /// Function pass entry point.
+  void runOnOperation() override;
+};
+
+struct LinalgVectorLoweringPass
+    : public LinalgVectorLoweringBase<LinalgVectorLoweringPass> {
+  LinalgVectorLoweringPass() = default;
+  LinalgVectorLoweringPass(const LinalgVectorLoweringPass &pass) {}
+
+  void runOnOperation() override;
+};
+}  // namespace
+
+/// Return the neutral element as a new Value.
+/// For now, just assume it is the zero of type.
+/// In the future, it should be the zero of type + op.
+static Value getNeutralOfLinalgOp(OpBuilder &b, OpOperand &op) {
+  auto t = getElementTypeOrSelf(op.get().getType());
+  return b.create<arith::ConstantOp>(op.getOwner()->getLoc(), t,
+                                     b.getZeroAttr(t));
+}
+
+/// Collect all Linalg ops, they must all have tensor semantics.
+/// For now this just fuses everything.
+// TODO: finer control.
+void LinalgFusePass::runOnOperation() {
+  FuncOp funcOp = getOperation();
+
+  // Set up tiling and vectorization options.
+  FailureOr<LinalgTilingAndFusionOptions> defaultTilingOptions =
+      getTileAndFuseOptionsFromConfig(funcOp, tilingLevel);
+  if (failed(defaultTilingOptions)) {
+    return signalPassFailure();
+  }
+  LinalgTilingAndFusionOptions tilingOptions = defaultTilingOptions.getValue();
+  bool doTiling = !tilingOptions.tileSizes.empty();
+  if (!tileSizes.empty()) {
+    doTiling = true;
+    tilingOptions.tileSizes = {tileSizes.begin(), tileSizes.end()};
+  }
```

Comment:
Maybe rewrite it as this way which makes it cleaner:

```cpp
bool doTiling = !tileSizes.empty();
if (doTiling) {
  tilingOptions.tileSizes = ...
}
```

Reply:
I am not sure that is the same logic. Tiling is needed if there were tile sizes recovered by looking at the `lowering.config` attributes. Then the `tileSizes` from command line, override that. Seems like what you have has a slightly different logic (I dont fully follow it actually).